### DISCUSSION TITLE
NEW KEY FEATURE: Point Cloud Customization

### DISCUSSION THREAD

**Discussion by Samahu** [ouster-lidar/ouster-ros#260] [Announcements]
Starting with v0.11.2 for ROS2 and v0.11.0 for ROS1 we have implemented a new feature that let's _ouster-ros_ users customize the PointCloud generated by the driver. This comes at a very little to no overhead on the driver operation. Here are some keys aspects to utilizing this feature:
* _ouster-ros_ driver users will now see a significant drop of the bandwidth (about 33%) when enabling the low data mode (`RNG15_RFL8_NIR8`) from the sensor and utilizing the `native` format -described below - instead of the previous implementation where the required bandwidth in the ROS driver will stay the same even if the sensor lidar udp profile has changed.
* Also, we are providing the users with the ability to work with the Velodyne format using Ouster sensors, or even use the bare minimum xyz fields.
* The feature will also work seamlessly for the DUAL mode which results in two point clouds produced. However, currently users can only apply the same PointCloud format to both point clouds (you can not choose one type for the first point cloud and a different one for the other topic).

User can start using this feature through the new launch file parameter `point_type` which is described below:

**point_type**: This parameter allows to customize the point cloud that the driver produces through its `/ouster/points` and `/ouster/points` topics. Choose one of the following values:
  - `original`: This uses the original point representation `ouster_ros::Point` of the ouster-ros driver.
  - `native`: directly maps all fields as published by the sensor to an equivalent point cloud representation with the additon of ring and timestamp fields.
  - `xyz`: the simplest point type, only has {x, y, z}
  - `xyzi`: same as `xyz` point type but adds intensity (signal) field.
  - `xyzir`: same as `xyzi` type but adds ring (channel) field. This type is same as Velodyne point cloud type.

> **_Note:_** Not that the PointCloud types that require the intensity field such as `xyzi` or `xyzir` are not compatible with Ouster's low data mode since it lacks signal (intensity) from its representation. If the user choose these point type along with the low data mode the driver will still operate but the intensity field will always bet set to zero.

> **_Note:_** starting with v0.11.0 for ROS1 the driver requires C++17 to compile.

All related issues and PRs:
- ROS2 #216, #254, #255
- ROS1 #253
- resolves #15 
- resolves #29
- resolves #210 
- resolves #97
- related #217
- related #223
- related #287

**Comment by James-R-Han** [comment]
Hi @Samahu!

Thanks for this great PR! One question I had was about the 't' (time) field of non-original type points. Is it possible to get something like txyz or txyzi? The time field is useful for point clouds to undistort them when the sensor is in motion, but I'd still like to get some bandwidth reduction.

Thank you in advance!

**Reply by Samahu** [reply]
I prototyped something similar to what you are asking for but I haven't merged it. You can check it out here: https://github.com/ouster-lidar/ouster-ros/pull/290/files. I would like to merge the PR at some point but in case you wanted to pick and choose which fields to include or exclude or add a derivative custom fields, then this PR shows you how to do that.

**Reply by GreatAlexander** [reply]
Hi @Samahu , thanks for all your great work on this repo. We are users of this Ouster driver and we are interested in the `xyzit` format as well (Not interested in Apollo/CyberRT support).
Are you aiming to merge that PR anytime soon, or should we just go ahead and create a fork with these changes?

**Reply by Samahu** [reply]
The issue is I wasn't sure how much widely this format is in use (I don't have a lot of users asking for it and at same time we have plenty of requests) .. I can try to merge it after I am doing with #363. I don't see a harm in doing so.
